File: frontends/web/src/lib/metrics.ts
import { formatMetricValue } from "@/lib/format";
import type { ChallengeDetailResponse } from "@/lib/schemas";

/** Describes the metric schema shape used by this module. */
type MetricSchema = ChallengeDetailResponse["spec"]["metric_schema"];
/** Describes the metric definition shape used by this module. */
type MetricDefinition = MetricSchema["metrics"][number];
/** Describes the metric value shape used by this module. */
type MetricValue = { metric_name: string; value: number };

/** Find display metadata for a metric name declared by the challenge bundle. */
export function metricDefinition(
  schema: MetricSchema,
  metricName: string,
): MetricDefinition | undefined {
  return schema.metrics.find((metric) => metric.name === metricName);
}

/** Return the aggregate metric selected as the primary ranking metric. */
export function primaryMetric(
  schema: MetricSchema,
  metrics: MetricValue[],
): MetricValue | undefined {
  return metrics.find(
    (metric) => metric.metric_name === schema.ranking.primary_metric_name,
  );
}

/** Human-facing metric label with a safe fallback for unknown names. */
export function metricLabel(schema: MetricSchema, metricName: string): string {
  return metricDefinition(schema, metricName)?.label ?? metricName;
}

/** Format a metric value using bundle metadata when available. */
export function formatDeclaredMetric(
  schema: MetricSchema,
  metric: MetricValue | undefined,
): string {
  if (!metric) return "n/a";
  return formatMetricValue(
    metric.value,
    metricDefinition(schema, metric.metric_name)?.unit,
  );
}

/** Compact direction label for leaderboard headers and metric metadata. */
export function metricDirectionLabel(
  direction: "maximize" | "minimize",
): string {
  return direction === "maximize" ? "higher is better" : "lower is better";
}

File: crates/domain/src/models/challenge/metrics.rs
use serde::{Deserialize, Serialize};

use super::super::names::MetricName;
use super::serde_helpers::{
    required_nullable, required_nullable_non_empty_vec, required_nullable_non_empty_vec_schema,
    required_nullable_schema, serialize_empty_vec_as_null,
};

/// Whether a metric is better when it is larger or smaller.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum MetricDirection {
    Maximize,
    Minimize,
}

/// Visibility level for a metric emitted by the evaluator.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum MetricVisibility {
    /// Visible in validation feedback and official result views.
    Public,
    /// Visible only after a ranking-visible official evaluation.
    Official,
}

/// One metric that an evaluator may emit in aggregate or per-run result payloads.
#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct MetricDefinitionSpec {
    pub name: MetricName,
    pub label: String,
    #[serde(deserialize_with = "required_nullable")]
    #[schemars(required, schema_with = "required_nullable_schema::<String>")]
    pub unit: Option<String>,
    pub direction: MetricDirection,
    pub visibility: MetricVisibility,
    #[serde(deserialize_with = "required_nullable")]
    #[schemars(required, schema_with = "required_nullable_schema::<String>")]
    pub metric_description: Option<String>,
}

/// Ranking configuration for a challenge.
#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct RankingSpec {
    pub primary_metric_name: MetricName,
    #[serde(
        deserialize_with = "required_nullable_non_empty_vec",
        serialize_with = "serialize_empty_vec_as_null"
    )]
    #[schemars(
        required,
        schema_with = "required_nullable_non_empty_vec_schema::<MetricName>"
    )]
    pub tie_breaker_metric_names: Vec<MetricName>,
}

/// Metric schema embedded in `spec.json`.
#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct MetricSchemaSpec {
    pub metrics: Vec<MetricDefinitionSpec>,
    pub ranking: RankingSpec,
}

impl MetricSchemaSpec {
    /// Look up a metric definition by name.
    pub fn metric(&self, metric_name: &MetricName) -> Option<&MetricDefinitionSpec> {
        self.metrics
            .iter()
            .find(|metric| &metric.name == metric_name)
    }

    /// Primary ranking metric declared by this challenge.
    pub fn primary_metric(&self) -> Option<&MetricDefinitionSpec> {
        self.metric(&self.ranking.primary_metric_name)
    }
}

impl Default for MetricSchemaSpec {
    /// Handles default for this module.
    fn default() -> Self {
        Self {
            metrics: vec![MetricDefinitionSpec {
                name: MetricName::score(),
                label: "Score".to_string(),
                unit: None,
                direction: MetricDirection::Maximize,
                visibility: MetricVisibility::Public,
                metric_description: Some("Challenge-defined compatibility score.".to_string()),
            }],
            ranking: RankingSpec {
                primary_metric_name: MetricName::score(),
                tie_breaker_metric_names: vec![],
            },
        }
    }
}
